### ISSUE TITLE
changing from goroutine that is reading a channel to another goroutine

### ISSUE THREAD

**Issue by ronaldpetty** [go-delve/delve#2665]
I am not 100% sure, but I think dlv stops taking input if the active goroutine is waiting on a channel. When I hit control-c it kills the app (not dlv). I was hoping that I could type 'grs' and pick another routine if the one in the dlv console was waiting, but I can't type anything aside control-c. My work around is set breakpoint on the channel read, and then change routines, but I hope there is something else I am unaware of to manage this.

I set a break on read, then let it go, but after it starts reading the channel, I see other goroutines output but can't change to them.

```
> k8s.io/apiserver/pkg/server.preparedGenericAPIServer.Run() /home/ubuntu/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/genericapiserver.go:402 (hits goroutine(1):1 total:1) (PC: 0x255ffea)
   397:			// Wait for all requests to finish, which are bounded by the RequestTimeout variable.
   398:			s.HandlerChainWaitGroup.Wait()
   399:		}()
   400:	
   401:		klog.V(1).Info("[graceful-termination] waiting for shutdown to be initiated")
=> 402:		<-stopCh
   403:	
   404:		// run shutdown hooks directly. This includes deregistering from the kubernetes endpoint in case of kube-apiserver.
   405:		err = s.RunPreShutdownHooks()
   406:		if err != nil {
   407:			return err
```

then let it continue

```
Breakpoint 7 (enabled) at 0x255ffea for k8s.io/apiserver/pkg/server.preparedGenericAPIServer.Run() /home/ubuntu/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/genericapiserver.go:402 (1)
(dlv) c
I0818 19:12:39.339624    1325 apf_controller.go:299] Starting API Priority and Fairness config controller
I0818 19:12:39.340054    1325 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0818 19:12:39.340187    1325 apiservice_controller.go:97] Starting APIServiceRegistrationController
I0818 19:12:39.340207    1325 cache.go:32] Waiting for caches to sync for APIServiceRegistrationController controller
I0818 19:12:39.340263    1325 controller.go:83] Starting OpenAPI AggregationController
I0818 19:12:39.340342    1325 customresource_discovery_controller.go:209] Starting DiscoveryController
I0818 19:12:39.340613    1325 available_controller.go:491] Starting AvailableConditionController
I0818 19:12:39.340634    1325 cache.go:32] Waiting for caches to sync for AvailableConditionController controller
I0818 19:12:39.341700    1325 cluster_authentication_trust_controller.go:440] Starting cluster_authentication_trust_controller controller
I0818 19:12:39.341721    1325 shared_informer.go:240] Waiting for caches to sync for cluster_authentication_trust_controller
I0818 19:12:39.341774    1325 autoregister_controller.go:141] Starting autoregister controller
I0818 19:12:39.341794    1325 cache.go:32] Waiting for caches to sync for autoregister controller
I0818 19:12:39.342297    1325 controller.go:132] OpenAPI AggregationController: action for item : Nothing (removed from the queue).
I0818 19:12:39.342408    1325 controller.go:132] OpenAPI AggregationController: action for item k8s_internal_local_delegation_chain_0000000000: Nothing (removed from the queue).
E0818 19:12:39.362391    1325 controller.go:152] Unable to remove old endpoints from kubernetes service: StorageError: key not found, Code: 1, Key: /registry/masterleases/172.31.43.6, ResourceVersion: 0, AdditionalErrorMsg: 
I0818 19:12:39.396085    1325 controller.go:85] Starting OpenAPI controller
I0818 19:12:39.396600    1325 naming_controller.go:291] Starting NamingConditionController
I0818 19:12:39.396730    1325 establishing_controller.go:76] Starting EstablishingController
I0818 19:12:39.396806    1325 nonstructuralschema_controller.go:192] Starting NonStructuralSchemaConditionController
I0818 19:12:39.397621    1325 apiapproval_controller.go:186] Starting KubernetesAPIApprovalPolicyConformantConditionController
I0818 19:12:39.397912    1325 crd_finalizer.go:266] Starting CRDFinalizer
I0818 19:12:39.426675    1325 dynamic_cafile_content.go:156] "Starting controller" name="client-ca-bundle::/home/ubuntu/ca.pem"
I0818 19:12:39.439279    1325 crdregistration_controller.go:111] Starting crd-autoregister controller
I0818 19:12:39.439314    1325 shared_informer.go:240] Waiting for caches to sync for crd-autoregister
I0818 19:12:43.734953    1325 storage_scheduling.go:109] all system priority classes are created successfully or already exist.
I0818 19:12:43.739383    1325 shared_informer.go:247] Caches are synced for crd-autoregister 
I0818 19:12:43.741326    1325 cache.go:39] Caches are synced for AvailableConditionController controller
I0818 19:12:43.741999    1325 cache.go:39] Caches are synced for APIServiceRegistrationController controller
I0818 19:12:43.744046    1325 cache.go:39] Caches are synced for autoregister controller
I0818 19:12:43.744370    1325 shared_informer.go:247] Caches are synced for cluster_authentication_trust_controller 
I0818 19:12:43.777875    1325 apf_controller.go:304] Running API Priority and Fairness config worker
```

I can no longer type in the dlv console (sans terminate app). Maybe thats expected, even so, be nice to have some confirmation of what I am trying. Thanks to all!

1. What version of Delve are you using (`dlv version`)?

```
./dlv version
Delve Debugger
Version: 1.7.1
Build: $Id: 3bde2354aafb5a4043fd59838842c4cd4a8b6f0b $
```

2. What version of Go are you using? (`go version`)?

```
/usr/local/go/bin/go version
go version go1.17 linux/amd64
```

3. What operating system and processor architecture are you using?

```
uname -a
Linux ip-172-31-43-6 5.4.0-1045-aws #47~18.04.1-Ubuntu SMP Tue Apr 13 15:58:14 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

4. What did you do?

Tried to change goroutine because it is stuck on reading a channel

5. What did you expect to see?

Hoped it would allow me to type in dlv console but the console is fronzen, however the app is working just fine

6. What did you see instead?

no input accepted (exepct control-c).

**Comment by ronaldpetty** [comment]
Here is a small example program.

```
package main

import (
	"time"
)

func main() {
	c := make(chan struct{})
	go func() {
		time.Sleep(time.Duration(100)*time.Second)
		c<-struct{}{}
	}()
	<-c
}
```

If you don't break on `<-c`, you can't change gr.

I am trying to debug a program (k8s apiserver) where you hit a channel and there are over 1000 to choose from so trying to find a technique to surf/control through that.

**Comment by aarzilli** [comment]
For the most part this is intended behavior. The `continue` command runs the program which means you can not inspect its memory. 
I'm not even sure what would mean to "change goroutine" as the program is running.
The part where ctrl-c kills the target program is weird, what should happen is that the program ends up suspended and delve's command prompt returns. I can not reproduce this part of the report with the example program, this is what I get:

```
$ dlv debug test.go
Type 'help' for list of commands.
(dlv) c
[here I press Ctrl-C]
received SIGINT, stopping process (will not forward signal)
> runtime.epollwait() /usr/local/go/src/runtime/sys_linux_amd64.s:666 (PC: 0x459b00)
Warning: debugging optimized function
   661:		MOVL	nev+16(FP), DX
   662:		MOVL	timeout+20(FP), R10
   663:		MOVQ	$0, R8
   664:		MOVL	$SYS_epoll_pwait, AX
   665:		SYSCALL
=> 666:		MOVL	AX, ret+24(FP)
   667:		RET
   668:	
   669:	// void runtime·closeonexec(int32 fd);
   670:	TEXT runtime·closeonexec(SB),NOSPLIT,$0
   671:		MOVL    fd+0(FP), DI  // fd
(dlv) 
```

To see what a program with many goroutines is doing I suggest taking a look at the options of the `goroutines` command.

**Comment by ronaldpetty** [comment]
@aarzilli thanks for the swift response. 

Agree on all points. I wasn't clear, control-c doesn't kill the dlv app, its stoping the app dlv is running. I also get the same output as you above.

I will close this. I had been looking (google/github) but wasn't finding anything about moving off a blocked operation (channel read) interactively, was hoping there was. guess for now, just don't debug that particular code :) and break elsewhere. The beauty of concurrency.
